contracts/oracle/src/contract/alarms/iter.rs

```diff
@@ -0,0 +1,133 @@
+use super::{MarketAlarms, PriceResult};
+use crate::ContractError;
+use finance::{
+    currency::{self, AnyVisitor, AnyVisitorResult, Currency},
+    price::base::BasePrice,
+};
+use marketprice::alarms::{AlarmsIterator, PriceAlarms};
+use sdk::cosmwasm_std::{Addr, Storage};
+use serde::{de::DeserializeOwned, Serialize};
+use swap::SwapGroup;
+
+struct AlarmsCmd<'a, 'b, OracleBase>
+where
+    OracleBase: Currency,
+{
+    storage: &'a dyn Storage,
+    price_alarms: &'static PriceAlarms<'static>,
+    price: &'b BasePrice<SwapGroup, OracleBase>,
+}
+
+impl<'a, 'b, OracleBase> AnyVisitor for AlarmsCmd<'a, 'b, OracleBase>
+where
+    OracleBase: Currency,
+{
+    type Error = ContractError;
+    type Output = AlarmsIterator<'a>;
+
+    fn on<C>(self) -> AnyVisitorResult<Self>
+    where
+        C: Currency + Serialize + DeserializeOwned,
+    {
+        Ok(self
+            .price_alarms
+            .alarms::<C, OracleBase>(self.storage, self.price.try_into()?))
+    }
+}
+
+/// Combines all alarms iterators, injecting price errors.
+pub struct AlarmsFlatten<'a, I> {
+    storage: &'a dyn Storage,
+    prices: I,
+    alarms: Option<AlarmsIterator<'a>>,
+}
+
+impl<'a, I> AlarmsFlatten<'a, I> {
+    pub fn new(storage: &'a dyn Storage, prices: I) -> Self {
+        Self {
+            storage,
+            prices,
+            alarms: None,
+        }
+    }
+
+    fn next_price<BaseC>(&mut self) -> Option<<Self as Iterator>::Item>
+    where
+        I: Iterator<Item = PriceResult<BaseC>> + 'a,
+        BaseC: Currency,
+    {
+        self.prices.find_map(|res_price| {
+            res_price
+                .and_then(|price| {
+                    currency::visit_any_on_ticker::<SwapGroup, _>(
+                        price.base_ticker(),
+                        AlarmsCmd {
+                            storage: self.storage,
+                            price_alarms: &MarketAlarms::PRICE_ALARMS,
+                            price: &price,
+                        },
+                    )
+                    .map(|mut alarms| {
+                        let alarm = alarms.next();
+                        self.alarms = Some(alarms);
+                        alarm.map(|res| res.map_err(ContractError::from))
+                    })
+                })
+                .unwrap_or_else(|err| Some(Err(err)))
+        })
+    }
+}
+
+impl<'a, I, BaseC> Iterator for AlarmsFlatten<'a, I>
+where
+    I: Iterator<Item = PriceResult<BaseC>> + 'a,
+    BaseC: Currency,
+{
+    type Item = Result<Addr, ContractError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.alarms
+            .as_mut()
+            .and_then(|alarms| alarms.next())
+            .map(|res| res.map_err(ContractError::from))
```

Comment:
Define a function `&mut AlarmsIterator -> Option<<Self as Iterator>::Item>` to reuse obtaining the next element and mapping the error. Then re-use that function in `fn next_price` lines 71-73.